src/minichain/serialization.py

```
@@ -0,0 +1,66 @@
+
+def _to_field_map(
+    value: Mapping[str, Any] | object, field_order: tuple[str, ...]
+) -> dict[str, Any]:
+    if isinstance(value, Mapping):
+        source = dict(value)
+    else:
+        source = {field: getattr(value, field) for field in field_order if hasattr(value, field)}
+
+    missing = [field for field in field_order if field not in source]
+    if missing:
+        raise ValueError(f"Missing required fields: {', '.join(missing)}")
```

Ruff (0.15.1)

[warning] Avoid specifying long messages outside the exception class

(TRY003)